File: Problems/String Manipulation Persistent Storage/KVStoreSingleFile.py
# Author: Tushar Jaiswal
# Creation Date: 2025-07-25

# # Implement a KVStore class that supports storing and retrieving key-value pairs, as well as persisting data to the filesystem and restoring it. The class should include the following methods:
# Things to note:
#     Both key and value are strings and may contain any characters, including newlines.
#     You must not use Python’s built-in serialization libraries (such as json).
#     The main focus of the problem is implementing your own serialization/deserialization methods.

# 2nd question How to write multiple files when a file cannot exceed one KB.

import pytest

class KVStore:
    FILENAME = "kvstore"
    
    def __init__(self):
        self.store = {}
        
    def getValue(self, key: str) -> str: 
        return self.store[key]
    
    def setValue(self, key: str, val: str):
        self.store[key] = val
        
    def clear(self):
        self.store.clear()
        
    def persist_to_disk(self):
        encoding = self.encode()
        
        with open(KVStore.FILENAME, "w") as file:
            file.write(encoding)
    
    def encode(self) -> str:
        encoding = ""
        
        for key, val in self.store.items():
            encoding += str(len(key)) + "." + key
            encoding += str(len(val)) + "." + val
            
        return encoding
        
    def restore_from_disk(self):
        with open(KVStore.FILENAME, "r") as file:
            encoding = file.read()
        self.decode(encoding)
            
    def decode(self, encoding: str):
        i = 0
        while i < len(encoding):
            delimiter_index = encoding.find(".", i)
            key_length = int(encoding[i:delimiter_index])
            i = delimiter_index + 1 + key_length
            key = encoding[delimiter_index + 1:i]
            
            delimiter_index = encoding.find(".", i)
            val_length = int(encoding[i:delimiter_index])
            i = delimiter_index + 1 + val_length
            val = encoding[delimiter_index + 1:i]
            
            self.store[key] = val
    
if __name__ == "__main__":
    kvstore = KVStore()
    
    kvstore.setValue("a", "1")
    assert kvstore.getValue("a") == "1"
    with pytest.raises(KeyError):kvstore.getValue("b")
    kvstore.setValue("bas", "123")
    kvstore.persist_to_disk()
    kvstore.clear()
    kvstore.restore_from_disk()
    assert kvstore.getValue("a") == "1"
    assert kvstore.getValue("bas") == "123"
    
    print("All tests passed")

File: Problems/String Manipulation Persistent Storage/KVStoreSingleFile.java
//  Author: Tushar Jaiswal
//  Creation Date: 2025-07-25

//  Implement a KVStore class that supports storing and retrieving key-value pairs, as well as persisting data to the filesystem and restoring it. The class should include the following methods:
//  Things to note:
//      Both key and value are strings and may contain any characters, including newlines.
//      You must not use Python’s built-in serialization libraries (such as json).
//      The main focus of the problem is implementing your own serialization/deserialization methods.

//  2nd question How to write multiple files when a file cannot exceed one KB.

import java.io.*;
import java.util.*;

public class KVStore {
    private static final String FILENAME = "kvstore";
    private Map<String, String> store;

    public KVStore() {
        store = new HashMap<>();
    }

    public String getValue(String key) {
        if (!store.containsKey(key)) {
            throw new NoSuchElementException("Key not found: " + key);
        }
        return store.get(key);
    }

    public void setValue(String key, String val) {
        store.put(key, val);
    }

    public void clear() {
        store.clear();
    }

    public void persistToDisk() throws IOException {
        String encoding = encode();
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(FILENAME))) {
            writer.write(encoding);
        }
    }

    private String encode() {
        StringBuilder encoding = new StringBuilder();
        for (Map.Entry<String, String> entry : store.entrySet()) {
            String key = entry.getKey();
            String val = entry.getValue();
            encoding.append(key.length()).append(".").append(key);
            encoding.append(val.length()).append(".").append(val);
        }
        return encoding.toString();
    }

    public void restoreFromDisk() throws IOException {
        StringBuilder encoding = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new FileReader(FILENAME))) {
            String line;
            while ((line = reader.readLine()) != null) {
                encoding.append(line);
            }
        }
        decode(encoding.toString());
    }

    private void decode(String encoding) {
        int i = 0;
        while (i < encoding.length()) {
            int delimIndex = encoding.indexOf(".", i);
            int keyLen = Integer.parseInt(encoding.substring(i, delimIndex));
            int keyStart = delimIndex + 1;
            int keyEnd = keyStart + keyLen;
            String key = encoding.substring(keyStart, keyEnd);

            i = keyEnd;
            delimIndex = encoding.indexOf(".", i);
            int valLen = Integer.parseInt(encoding.substring(i, delimIndex));
            int valStart = delimIndex + 1;
            int valEnd = valStart + valLen;
            String val = encoding.substring(valStart, valEnd);

            store.put(key, val);
            i = valEnd;
        }
    }

    public static void main(String[] args) throws IOException {
        KVStore kvstore = new KVStore();

        // Set and Get
        kvstore.setValue("a", "1");
        assert kvstore.getValue("a").equals("1");

        // Exception check (simulate KeyError)
        try {
            kvstore.getValue("b");
            throw new AssertionError("Expected exception not thrown for key 'b'");
        } catch (NoSuchElementException e) {
            // Expected
        }

        kvstore.setValue("bas", "123");
        kvstore.persistToDisk();
        kvstore.clear();
        kvstore.restoreFromDisk();

        assert kvstore.getValue("a").equals("1");
        assert kvstore.getValue("bas").equals("123");

        System.out.println("All tests passed");
    }
}
